src/service.rs

```diff
@@ -0,0 +1,41 @@
+use std::sync::mpsc::channel;
+use std::sync::mpsc::{Receiver, Sender};
+
+pub struct ServiceInstance<O> {
+    pub receiver: Option<Receiver<O>>,
+    sender: Sender<O>,
+}
+
+impl<O> ServiceInstance<O> {
+    pub fn new() -> ServiceInstance<O> {
+        let (sender, receiver) = channel();
+        ServiceInstance {
+            receiver: Some(receiver),
+            sender
+        }
+    }
+
+    pub fn sender(&self) -> Sender<O> {
+        self.sender.clone()
+    }
+
+    pub fn receiver(&mut self) -> Receiver<O> {
+        match self.receiver.take() {
+            Some(receiver) => { receiver }
+            _ => { panic!("failed to extract receiver from service- is it already on?"); }
+        }
+    }
+}
+
+macro_rules! define_services {
+    ($( (module: $service:path, name: $service_instance:ident, dependencies: [$($dependency:ident),*])),*) => (
+        $(let mut $service_instance = ServiceInstance::new();)*
+        $(
+            paste::expr! {
+                $(let [<$dependency _clone>] = $dependency.sender(););*
+                let receiver = $service_instance.receiver();
+                thread::spawn(move || $service(receiver, $({[<$dependency _clone>]}),*));
+            }
+        )*
+    )
+}
```

Comment:
I can't stop writing macros. This macro just replaces what used to be written out by hand in main. Get a clone of all the receivers for the services it depends on, then call the service method in a new thread

I had to add the hacky paste library so that we could create new identifiers- we want to do this:
```
let messenger_clone = messenger.sender();
```
with just messenger as input (since there's no reason for the caller of the macro to input or know about the name for this variable)
Turns this out this is totally impossible in vanilla rust which blows. So I'm using this library that somehow does it with the magic of procmacros.

I've added a comment into the dependencies saying that we can :fire: that dependency if it ever causes us problems later and just go back to the old ugly way of defining and intertwining servies